src/timer.rs

```diff
@@ -0,0 +1,384 @@
+  ```rust
+    let timers = dp.TIMER.split();
+
+    let ch0 = timers
+        .channel0
+        .set_clock_source(ClockSource::Clock1Khz, 1_000u32.Hz());
+
+    ch0.enable_match1_interrupt();
+    ch0.enable_match1_interrupt();
+    ch0.disable_match2_interrupt();
+
+    ch0.set_preload_value(Milliseconds::new(0));
+    ch0.set_preload(hal::timer::Preload::PreloadMatchComparator1);
+    ch0.set_match0(Milliseconds::new(3_000u32));
+    ch0.set_match1(Milliseconds::new(7_000u32));
+
+    ch0.enable(); // start timer
+  ```
+*/
+
+use crate::{clock::Clocks, pac};
+use bl602_pac::TIMER;
+use embedded_time::{
+    duration::Milliseconds,
+    rate::{Extensions, Hertz},
+};
+use paste::paste;
+
+/// Error for [CountDown](embedded_hal::timer::CountDown)
+#[derive(Debug)]
+pub enum CountDownError {
+    /// Indicates that the clock wrapped during count down
+    Wrapped,
+}
+
+/// Clock sources for a timer channel.
+/// There are four timer clock sources available.
+pub enum ClockSource<'a> {
+    /// System master clock
+    Fclk(&'a Clocks),
+    /// 32K clock
+    Rc32Khz,
+    /// 1K clock (32K frequency division)
+    Clock1Khz,
+    /// 32M clock
+    Pll32Mhz,
+}
+
+impl<'a> ClockSource<'a> {
+    fn tccr_value(&self) -> u8 {
+        match self {
+            ClockSource::Fclk(_) => 0,
+            ClockSource::Rc32Khz => 1,
+            ClockSource::Clock1Khz => 2,
+            ClockSource::Pll32Mhz => 3,
+        }
+    }
+
+    fn hertz(&self) -> Hertz {
+        match self {
+            ClockSource::Fclk(clocks) => clocks.sysclk(),
+            ClockSource::Rc32Khz => 32_000.Hz(),
+            ClockSource::Clock1Khz => 1_000.Hz(),
+            ClockSource::Pll32Mhz => 32_000_000.Hz(),
+        }
+    }
+}
+
+/// When to preload
+pub enum Preload {
+    /// No preload
+    NoPreload,
+    /// Preload when comparator 0 matches
+    PreloadMatchComparator0,
+    /// Preload when comparator 1 matches
+    PreloadMatchComparator1,
+    /// Preload when comparator 2 matches
+    PreloadMatchComparator2,
+}
+
+impl Preload {
+    fn to_prlcr(&self) -> u8 {
+        match self {
+            Preload::NoPreload => 0,
+            Preload::PreloadMatchComparator0 => 1,
+            Preload::PreloadMatchComparator1 => 2,
+            Preload::PreloadMatchComparator2 => 3,
+        }
+    }
+}
+
+/// Timer Channel 0
+pub struct TimerChannel0 {}
+
+/// Timer Channel 1
+pub struct TimerChannel1 {}
+
+/// Timers obtained from [TIMER.split](bl602_pac::Peripherals::TIMER)
+pub struct Timers {
+    pub channel0: TimerChannel0,
+    pub channel1: TimerChannel1,
+}
+
+macro_rules! impl_timer_channel {
+    ($name: ident, $conf_name: ident, $channel: literal, $channel_cs: literal) => {
+
+        /// A configured timer channel ready to use.
+        ///
+        /// Be cautios when creating a [Timer](embedded_time::Timer) from this and using `wait` on it.
+        /// If the timer is configured to never reach the target time it will block forever.
```

Comment:
Is it possible to catch an invalid configuration at compile time, or panic at invocation time?

Reply:
it now remembers the last read value and if the timer wrapped it now generates an error which the user can check

Reply:
oh wait ... that was a bad idea since it makes a simple `try_now` error if the time wrapped which isn't a good idea
maybe it just shouldn't implement embedded_time::Clock at all

Reply:
I decided to remove implementation of embedded_time::Clock for now since I have no idea how to implement it in a sane way.